controlplane/kubeadm/controllers/helpers.go

```diff
@@ -224,8 +225,30 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 		},
 	}
 
+	if err := setInfraRefAnnotationToMachine(kcp.Spec.InfrastructureTemplate, machine); err != nil {
+		return errors.Wrap(err, "Failed to set infrastructure-template-reference annotation")
+	}
```

Comment:
We should set the annotation in the external clone function in the general case

Reply:
We can add this where machine is created. In the clone function, infrastructure object is being created not the machine. Am I missing something? 
So, there is another place where machine is created in MachineSet: https://github.com/kubernetes-sigs/cluster-api/blob/714b0a97484465a05a67d8ccd1b64aa388b9fac2/controllers/machineset_controller.go#L357

Reply:
And machine deployment already adds template hash as label: DefaultMachineDeploymentUniqueLabelKey

Reply:
Apologies, I meant the annotation should be on the cloned object, not on the Machine itself.  In https://github.com/kubernetes-sigs/cluster-api/blob/714b0a97484465a05a67d8ccd1b64aa388b9fac2/controllers/external/util.go#L79 we have all the information we need to set the reference generically.

Reply:
So this hash will be used by CAPI controllers, not by infrastructure provider controllers. And the cloned template is there in the spec for infra object (AWSMachine), why also keep the cloned template hash? When there is a change in AWSMachine spec, it will be upgraded, no need to check the hash. WDYT?

Reply:
Also, if we think about where we are going to use it: in kubeadm control-plane controller, we want to annotate machines with cloned infra-template annotation if they are missing it (either adopted machines or older version machines). 

It may make things slow IMO if at every reconcile we go and fetch infra object.

Reply:
Just to clarify, the value of the cloned-from isn't a hash but a json representation of a ObjectReference. This information might be used in both CAPI and other controllers (e.g. @sethp-nr wants to use it in CAPA).

> It may make things slow IMO if at every reconcile we go and fetch infra object.
Because we use a cached delegating client, `Get` requests on infrastructure objects will be cached as well.

Reply:
> So this hash will be used by CAPI controllers, not by infrastructure provider controllers.

It's both: the original reason to have the link was so CAPA could reconcile changes from the template down to the AWSMachine instances cloned from that template without needing the full machine replacement flow.

Reply:
@vincepri @sethp-nr  Thanks for the comments. Added the annotation to infra machine. PTAL when you get a chance.